.emacs.d/config.org
#+TITLE: ID's GNU Emacs Configuration File
#+AUTHOR: CToID
#+OPTIONS: num:nil

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

*This config file is no longer used*

Check [[file:init.org][the new config]] instead

* Table of Contents                                            :TOC:noexport:
- [[#configure-package-and-use-package][Configure Package and use-package]]
- [[#appearance-packages-setup][Appearance Packages Setup]]
  - [[#themes-and-icons][Themes and Icons]]
  - [[#startup-screen-dashboard][Startup Screen (Dashboard)]]
- [[#org-mode][Org Mode]]
  - [[#toc-org][toc-org]]
  - [[#org-present][org-present]]
- [[#other-packages-setup][Other Packages Setup]]
  - [[#magit][magit]]
  - [[#browse-kill-ring][browse-kill-ring]]
  - [[#helpful][helpful]]
  - [[#xclip][xclip]]
  - [[#vterm][vterm]]
  - [[#which-key][which-key]]
  - [[#neotree][neotree]]
  - [[#recentf][recentf]]
  - [[#sudo-edit][sudo-edit]]
  - [[#highlight-indent-guides][highlight-indent-guides]]
  - [[#visual-line-mode][visual-line-mode]]
  - [[#emms][EMMS]]
  - [[#treesit-auto][treesit-auto]]
- [[#languages][Languages]]
  - [[#arduino][Arduino]]
  - [[#yaml][YAML]]
  - [[#common-lisp][Common Lisp]]
  - [[#racket][Racket]]
  - [[#markdown][Markdown]]
- [[#hooks][Hooks]]
  - [[#startup-hook][Startup Hook]]
  - [[#server-hook][Server Hook]]
  - [[#hl-line-mode-hook][hl-line-mode-hook]]
- [[#functions][Functions]]
  - [[#toggle-transparency][toggle-transparency]]
  - [[#font-and-face-settings][Font and Face Settings]]
  - [[#toggle-buffer-visibility-in-ibuffer][Toggle Buffer Visibility in ibuffer]]
  - [[#set-keys][Set Keys]]
  - [[#vterm-frame][VTerm Frame]]
  - [[#eshell-frame][Eshell Frame]]
  - [[#dired-frame][Dired Frame]]
  - [[#emms-mpv-volume-functions][EMMS MPV volume functions]]

* Configure Package and use-package
Add package archives and make sure use-package is installed.
#+begin_src elisp
(require 'package)
(add-to-list 'package-archives
             '("melpa" . "https://melpa.org/packages/")
             '("gnu" . "https://elpa.gnu.org/packages/"))

(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))

(eval-when-compile
  ;; (add-to-list 'load-path "/home/id/.emacs.d/elpa")
  (require 'use-package))
;; (setq use-package-verbose t)

(setq package--init-file-ensured t)
#+end_src

* COMMENT Gcmh Setup
the Garbage Collector Magic Hack package will adjust garbage collection interference, thus speeding up Emacs startup process.
#+begin_src elisp
(use-package gcmh
  :ensure t
  :config
  (gcmh-mode 1))
#+end_src

* Appearance Packages Setup
** Themes and Icons
Using Challenger Deep theme.
#+begin_src elisp
(use-package doom-themes
  :ensure t
  :init
  (setq doom-themes-enable-bold t
        doom-challenger-deep-brighter-modeline t
        doom-themes-enable-italic t)
  :config
  (load-theme 'doom-challenger-deep t))

(use-package all-the-icons
  :ensure t
  ;; :if (display-graphic-p)
  :config
  (unless all-the-icons-font-names
    (all-the-icons-install-fonts)))

(use-package all-the-icons-dired
  :ensure t
  :after all-the-icons
  :hook (dired-mode . all-the-icons-dired-mode))

(use-package all-the-icons-ibuffer
  :ensure t
  :after all-the-icons
  :hook (ibuffer-mode  . all-the-icons-ibuffer-mode))

;; (use-package doom-modeline
;;   :ensure t
;;   :init
;;   (doom-modeline-mode 1))
#+end_src

** Startup Screen (Dashboard)
#+begin_src elisp
(use-package dashboard
  :ensure t
  :init
  (setq dashboard-banner-logo-title "GNU EMACS - Editor MACroS"
        ;; dashboard-startup-banner (nth (% (random most-positive-fixnum) 6) '(official logo 1 2 3 "~/.emacs.d/emacs.png"))
        dashboard-startup-banner "~/.emacs.d/emacs.png"
        dashboard-center-content t
        dashboard-week-agenda nil
        dashboard-agenda-time-string-format "%Y-%m-%d %H:%M"
        dashboard-agenda-prefix-format "%?-12t% s"
        dashboard-agenda-release-buffers t
        dashboard-items nil
        dashboard-set-init-info nil)
  :config
  (add-to-list 'dashboard-footer-messages "Free as in Freedom!")
  (add-to-list 'dashboard-footer-messages "There is no system but GNU, and Linux is one of its kernels.")
  ;; This is to ensure dashboard still spawn if using "emacs -f server-start" to start an emacs server
  (if (and (< (length command-line-args) 4)
           (member "server-start" command-line-args))
      (let ((command-line-args '("emacs")))
        (dashboard-setup-startup-hook))
    (dashboard-setup-startup-hook)))
#+end_src

* COMMENT Key bindings packages setup
** General
For binding leader keys.
#+begin_src elisp
(use-package general
  :ensure t
  :config
  (general-evil-setup t))
#+end_src

** Evil Mode
An extensible "vim" layer for Emacs.
#+begin_src elisp
(use-package evil
  :ensure t
  :init
  (setq evil-want-C-u-scroll t
        evil-want-integration t
        evil-want-keybinding nil
        evil-want-minibuffer t)
  :config
  (evil-mode 1))
#+end_src

*** evil-collection
Evil mode extesions for many other modes
#+begin_src elisp
(use-package evil-collection
  :after evil
  :ensure t
  :config
  (evil-collection-init))
#+end_src

*** evil-mc
Multi cursors implementation for evil-mode
#+begin_src elisp
(use-package evil-mc
  :commands (evil-mc-mode) 
  :ensure t)
#+end_src

* Org Mode
** toc-org
For creating table of contents in Org mode.
#+begin_src elisp 
(use-package toc-org
  :commands toc-org-enable
  :ensure t
  :hook (org-mode . toc-org-enable))
#+end_src

** org-present
For presentation in org-mode.
#+begin_src elisp
(use-package org-present
  :commands org-present
  :bind (:map org-present-mode-keymap
              ("C-c C-t" . global-display-line-numbers-mode))
  :ensure t
  :config
  (add-hook 'org-present-mode-hook
            (lambda ()
              (setq org-image-actual-width 1000
                    header-line-format " ")
              (setq-local face-remapping-alist '((default (:height 2.0) default)
                                                 (header-line (:height 5.0) header-line)
                                                 (org-document-title (:height 2.25) org-document-title)
                                                 (org-code (:height 1.0) org-code)
                                                 (org-verbatim (:height 1.0) org-verbatim)
                                                 (org-block (:height 1.0) org-block)
                                                 (org-block-begin-line (:height 0.75) org-block-begin-line)))
              ;; (set-fontset-font "fontset-default" 'han "Noto Sans TC-20")
              ;; (set-fontset-font "fontset-default" 'han "Noto Sans SC-20"
              ;;                   nil 'append)
              ;; (set-fontset-font "fontset-default" 'big5 "Noto Sans TC-20")
              (global-display-line-numbers-mode -1)
              ;; (org-present-big)
              (org-display-inline-images)
              (org-present-read-only)
              (visual-fill-column-mode 1)
              (org-present-hide-cursor)
              (visual-line-mode 1)))
  (add-hook 'org-present-mode-quit-hook
            (lambda ()
              (setq org-image-actual-width 300
                    header-line-format nil)
              (setq-local face-remapping-alist '((default (:height 1.0) default)))
              ;; (set-fontset-font "fontset-default" 'han "Noto Sans TC-12")
              ;; (set-fontset-font "fontset-default" 'han "Noto Sans SC-12"
              ;;                   nil 'append)
              ;; (set-fontset-font "fontset-default" 'big5 "Noto Sans TC-12")
              (global-display-line-numbers-mode 1)
              ;; (org-present-small)
              (org-remove-inline-images)
              (org-present-read-write)
              (visual-fill-column-mode 0)
              (org-present-show-cursor)
              (visual-line-mode 0))))
#+end_src

* Other Packages Setup
** magit
A git client for emacs.
#+begin_src elisp
(use-package magit
  :ensure t
  :commands (magit magit-clone))
#+end_src

** browse-kill-ring
A nice little kill ring browser.
#+begin_src elisp
(use-package browse-kill-ring
  :ensure t
  :config
  (browse-kill-ring-default-keybindings)
  (setq browse-kill-ring-depropertize t
        browse-kill-ring-display-duplicates t))
#+end_src

** helpful
A better help buffer.
#+begin_src elisp
(use-package helpful
  :ensure t)
#+end_src

** xclip
#+begin_src elisp
(use-package xclip
  :ensure t)
#+end_src

** vterm
Terminal emulator.
#+begin_src elisp
(use-package vterm
  :commands vterm
  :ensure t)
#+end_src

** COMMENT undo-tree
An undo system to work with Evil.
#+begin_src elisp
(use-package undo-tree
  :after evil
  :ensure t
  :init
  (setq undo-tree-history-directory-alist '(("." . "~/.emacs.d/undo_tree")))
  :config
  (evil-set-undo-system 'undo-tree)
  (global-undo-tree-mode 1))
#+end_src

** which-key
Minor mode for showing following key bindings based on your currently entered incomplete command.
#+begin_src elisp
(use-package which-key
  :ensure t
  :init
  (setq which-key-side-window-location 'bottom
        which-key-sort-uppercase-first nil
        which-key-add-column-padding 1
        which-key-max-display-columns nil
        which-key-side-window-max-height 0.25
        which-key-idle-delay 0.5
        which-key-max-description-length 25)
  :config
  (which-key-mode))
#+end_src

** neotree
A file tree viewer.
#+begin_src elisp
(use-package neotree
  :commands (neotree-toggle neotree-dir)
  :ensure t
  :config
  (setq neo-smart-open t
        neo-window-width 30
        neo-window-fixed-size nil
        neo-theme (if (display-graphic-p) 'icons 'arrow)
        inhibit-compacting-font-caches t))
#+end_src

** recentf
For showing recently open files.
#+begin_src elisp
(use-package recentf
  :init
  (setq recentf-max-menu-items 10
        recentf-max-saved-items 10)
  (add-to-list 'recentf-exclude "\\.last\\'")
  :config
  (recentf-mode 1))
#+end_src

** sudo-edit
For editing a file in sudo.
#+begin_src elisp
(use-package sudo-edit
  :commands (sudo-edit)
  :ensure t
  :config
  (sudo-edit-indicator-mode 1))
#+end_src

** highlight-indent-guides
For highlighting indentation levels.
#+begin_src elisp
(use-package highlight-indent-guides
  :commands (highlight-indent-guides-mode)
  :ensure t
  :config
  (setq highlight-indent-guides-method 'column
        highlight-indent-guides-responsive 'top))
#+end_src

** visual-line-mode
A package to make org-present looks better.
#+begin_src elisp
(use-package visual-fill-column
  :after org-present
  :ensure t
  :init
  (setq visual-fill-column-width 225
        visual-fill-column-center-text t))
#+end_src

** EMMS
Just EMMS.
#+begin_src elisp
(use-package emms
  :ensure t
  :config
  (emms-all)
  (setq emms-player-list '(emms-player-mpv)
        emms-info-functions '(emms-info-native)
        emms-volume-change-function #'emms-volume-mpv-change)
  :bind (("C-c m p" . emms-pause)
         ("C-c m +" . emms-volume-mode-plus)
         ("C-c m -" . emms-volume-mode-minus)))
#+end_src

** treesit-auto
A package for easier use of tree sitter library.
#+begin_src elisp
(use-package treesit-auto
  :ensure t
  :init
  :config
  (setq treesit-auto-install 'prompt)
  (global-treesit-auto-mode))
#+end_src

* Languages
** COMMENT LSP
*** Eglot
#+begin_src elisp
(use-package eglot
  :config
  (add-hook 'haskell-mode-hook 'eglot-ensure)
  (setq-default eglot-workspace-configuration
                '((haskell (plugin (stan (globalOn . :json-false))))))
  :custom
  (eglot-autoshutdown t)
  (eglot-confire-server-initiated-edits nil))
#+end_src

** Arduino
*** arduino-mode
#+begin_src elisp
(use-package arduino-mode
  :mode "\\.ino\\'"
  :ensure t)
#+end_src

** COMMENT Rust
*** rust-mode
#+begin_src elisp
(use-package rust-mode
  :mode "\\.rs\\'"
  :ensure t)
#+end_src

** COMMENT VimScript
*** vimrc-mode
#+begin_src elisp
(use-package vimrc-mode
  :mode "\\.vim\\'"
  :ensure t)
#+end_src

** YAML
*** yaml-mode
#+begin_src elisp
(use-package yaml-mode
  :mode "\\.yml\\'"
  :ensure t)
#+end_src

** Common Lisp
*** Sly
newer SLIME.
#+begin_src elisp
(use-package sly
  :commands (sly)
  :ensure t
  :init
  (setq inferior-lisp-program "sbcl"))
#+end_src
*** COMMENT SLIME
For Common Lisp.
#+begin_src elisp
(use-package slime
  :commands (slime)
  :ensure t
  :init
  (setq inferior-lisp-program "sbcl"))
#+end_src

** Racket
*** racket-mode
#+begin_src elisp
(use-package racket-mode
  :ensure t
  :mode "\\.rkt\\'")
#+end_src

** Markdown
*** markdown-mode
#+begin_src elisp
(use-package markdown-mode
  :mode "\\.md\\'")
#+end_src

** COMMENT Haskell
*** haskell-mode
#+begin_src elisp
(use-package haskell-mode
  :ensure t
  :mode "\\.hs\\'")
#+end_src
* COMMENT Key Bindings Setup
Use these with Evil mode.
** Leader key
#+begin_src elisp
(nvmap :prefix "SPC"
#+end_src

** Buffer controls
#+begin_src elisp
  "b c" '(kill-current-buffer :which-key "Close current buffer")
  "b C" '(kill-buffer :which-key "Close buffer")
  "b h" '(previous-buffer :which-key "Previous buffer")
  "b l" '(next-buffer :which-key "Next buffer")
  "b s" '(switch-to-buffer :which-key "Switch to buffer")
  "b r" '(revert-buffer :whitch_key "Revert buffer")
#+end_src

** Window controls
#+begin_src elisp
  "w c" '(delete-window :which-key "Close current window")
  "w C" '(delete-other-windows :which-key "Close other windows")
  "w s" '(split-window-right :which-key "Split window Horizontally")
  "w v" '(split-window-below :which-key "Split window Vertically")
  "w h" '(windmove-left :which-key "Window left")
  "w j" '(windmove-down :which-key "Window down")
  "w k" '(windmove-up :which-key "Window up")
  "w l" '(windmove-right :which-key "Window right")
  "w b" '(balance-windows :which-key "Balance windows")
  "w H" '(shrink-window-horizontally :which-key "Shrink horizontal window")
  "w J" '(enlarge-window :which-key "Enlarge window")
  "w K" '(shrink-window :which-key "Shrink vertical window")
  "w L" '(enlarge-window-horizontally :which-key "Enlarge horizontal window")
#+end_src

** File actions
#+begin_src elisp
  "f r" '(recentf-open-files :which-key "Open recent file list")
  "f s" '(sudo-edit :which-key "Open/Edit file as sudo")
  "n t" '(neotree-toggle :which-key "Toggle neotree file viewer")
  "n d" '(neotree-dir :whick-key "Open directory in neotree")
#+end_src

** Org actions
#+begin_src elisp
  ;; "o s" '(org-insert-structure-template :which-key "Insert Org structure")
  "o t c" '(org-table-create :which-key "Create Org table")
  "o a" '(org-agenda :which-key "Open Org agenda view")
  "o t t" '(org-table-toggle-coordinate-overlays :which-key "Toggle Org table coordinates")
#+end_src

** Emms actions
#+begin_src elisp
  "m p" '(emms-pause :which-key "Play/Pause emms")
  "m m" '(emms :which-key "Open emms")
  "m f" '(emms-seek-forward :which-key "Emms seek forward")
  "m b" '(emms-seek-backward :which-key "Emms seek backward")
  "m -" '(emms-player-mpv-lower-volume :which-key "Lower volume")
  "m =" '(emms-player-mpv-raise-volume :which-key "Raise volume")
  "m r" '(emms-toggle-repeat-track :which-key "Toggle repeat track")
  "m R" '(emms-toggle-repeat-playlist :which-key "Toggle repeat playlist")
  "m l" '(emms-next :which-key "Next track")
  "m h" '(emms-previous :which-key "Previous track")
  "m a u" '(emms-add-url :which-key "Add URL to playlist")
  "m a f" '(emms-add-file :which-key "Add a file to playlist")
  "m t v" '(emms-mpv-toggle-video :which-key "Toggle to play video or not")
#+end_src

** Emacs actions
#+begin_src elisp
  "E r c" '((lambda () (interactive)
                (load-file "~/.emacs.d/init.el"))
              :which-key "Reload emacs config")
  "E t t" '(toggle-truncate-lines :which-key "Toggle truncate lines")
  "E t m" '(menu-bar-mode :which-key "Toggle menu bar")
  "E t c" '(evil-mc-mode :which-key "Toggle Multi cursors mode")
  "E t g" '(highlight-indent-guides-mode :which-key "Toggle indentation guides")
  "E t w" '(whitespace-mode :which-key "Toggle whitespace mode"))
#+end_src

* Hooks
** Startup Hook
#+begin_src elisp
(add-hook 'emacs-startup-hook
          (lambda ()
            (require 'ibuf-ext)
            (setq file-name-handler-alist temp-alist
                  frame-title-format "%b - GNU Emacs"
                  elfeed-db nil
                  gc-cons-threshold (* 32 1024 1024)
                  gc-cons-percentage 0.25
                  default-input-method "chinese-array30"
                  org-default-notes-file (locate-user-emacs-file "notes.org")
                  initial-buffer-choice (lambda () (get-buffer-create "*dashboard*")))
            (setq-default python-indent-offset 2
                          rust-indent-offset 2
                          css-indent-offset 2
                          js-indent-level 2
                          standard-indent 2
                          perl-indent-level 2)
            (defalias 'perl-mode 'cperl-mode)
            ;; (add-to-list 'ibuffer-never-show-predicates "^\\*")
            (add-hook 'org-mode-hook 'org-indent-mode)
            (add-hook 'ibuffer-mode-hook
                      (lambda ()
                        (ibuffer-switch-to-saved-filter-groups "customized")))
            (add-hook 'image-mode-new-window-functions
                      (lambda (arg)
                        (display-line-numbers-mode -1)))
            (add-hook 'xwidget-webkit-mode-hook
                      (lambda ()
                        (display-line-numbers-mode -1)))
            (ido-mode t)
            (set-default-coding-systems 'utf-8)
            (set-fonts)
            (set-keys)
            (package-initialize)))
#+end_src

** Server Hook
Set fonts again after spawning a frame.
#+begin_src elisp
(add-hook 'server-after-make-frame-hook
          (lambda ()
            (set-fonts)
            (if (equal (buffer-name) "*dashboard*")
                (revert-buffer))))
#+end_src

** hl-line-mode-hook
#+begin_src elisp
(add-hook 'hl-line-mode-hook
          (lambda ()
            (if hl-line-mode
                (set-face-attribute 'line-number-current-line nil
                                    :inherit 'hl-line)
              (set-face-attribute 'line-number-current-line nil
                                  :inherit 'default))))
#+end_src

* Functions
** toggle-transparency
#+begin_src elisp
(defun toggle-transparency ()
  (interactive)
  (let ((alpha (frame-parameter nil 'alpha)))
    (set-frame-parameter
     nil 'alpha
     (if (eql (cond ((numberp alpha) alpha)
                    ((numberp (cdr alpha)) (cdr alpha))
                    ((numberp (cadr alpha)) (cadr alpha)))
              100)
         '(90 . 75) '(100 . 100)))))
#+end_src

** Font and Face Settings
#+begin_src elisp
(defun set-fonts ()
  (set-fontset-font "fontset-default" 'han (font-spec :family "HanaMinA"))
  (set-fontset-font "fontset-default" 'han (font-spec :family "HanaMinB")
                    nil 'append)

.config/qtile/config.py
# Copyright (c) 2010 Aldo Cortesi
# Copyright (c) 2010, 2014 dequis
# Copyright (c) 2012 Randall Ma
# Copyright (c) 2012-2014 Tycho Andersen
# Copyright (c) 2012 Craig Barnes
# Copyright (c) 2013 horsik
# Copyright (c) 2013 Tao Sauvage
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import subprocess, os, traverse
from libqtile import bar, layout, widget, extension
from libqtile.config import Click, Drag, Group, Key, Match, Screen, hook, ScratchPad, DropDown, KeyChord
from libqtile.lazy import lazy
from libqtile.utils import guess_terminal

mod = "mod4"
terminal = guess_terminal()

keys = [
    # A list of available commands that can be bound to keys can be found
    # at https://docs.qtile.org/en/latest/manual/config/lazy.html
    # Switch between windows
    Key ([mod], "h", lazy.layout.left(), desc="Move focus to left"),
    Key ([mod], "l", lazy.layout.right(), desc="Move focus to right"),
    Key ([mod], "j", lazy.layout.down(), desc="Move focus down"),
    Key ([mod], "k", lazy.layout.up(), desc="Move focus up"),
    Key ([mod], "space", lazy.layout.next(), desc="Move window focus to other window"),
    # Move windows between left/right columns or move up/down in current stack.
    # Moving out of range in Columns layout will create new column.
    Key ([mod, "shift"], "h", lazy.layout.shuffle_left(), desc="Move window to the left"),
    Key ([mod, "shift"], "l", lazy.layout.shuffle_right(), desc="Move window to the right"),
    Key ([mod, "shift"], "j", lazy.layout.shuffle_down(), desc="Move window down"),
    Key ([mod, "shift"], "k", lazy.layout.shuffle_up(), desc="Move window up"),
    # Grow windows. If current window is on the edge of screen and direction
    # will be to screen edge - window would shrink.
    KeyChord ([mod, "control"], "r", [Key ([], "h", lazy.layout.grow_left (), desc = "Grow window to the left"),
                                      Key ([], "j", lazy.layout.grow_down (), desc = "Grow window down"),
                                      Key ([], "k", lazy.layout.grow_up (), desc = "Grow window up"),
                                      Key ([], "l", lazy.layout.grow_right (), desc = "Grow window to the right"),
                                      Key (["shift"], "6", lazy.layout.increase_ratio (), desc = "Grow the main window"),
                                      Key (["shift"], "v", lazy.layout.decrease_ratio (), desc = "Shrink the main window")],
              mode = True, name = "Resize Mode"),
    Key ([mod], "n", lazy.layout.normalize(), desc="Reset all window sizes"),
    # Toggle between split and unsplit sides of stack.
    # Split = all windows displayed
    # Unsplit = 1 window displayed, like Max layout, but still with
    # multiple stack panes
    Key (
        [mod, "shift"],
        "Return",
        lazy.layout.toggle_split(),
        desc="Toggle between split and unsplit sides of stack",
    ),
    Key ([mod], "Return", lazy.spawn(terminal), desc="Launch terminal"),
    # Emacs related keys
    KeyChord ([mod], "e", [Key ([], "e", lazy.spawn ("emacsclient -c -a ''"), desc = "Launch Emacs"),
                           Key ([], "d", lazy.spawn ("emacsclient -c -e '(make-dired-frame)'"), desc = "Launch Dired"),
                           Key ([], "s", lazy.spawn ("emacsclient -c -e '(make-eshell-frame)'"), desc = "Launch Eshell"),
                           Key ([], "Return", lazy.spawn ("emacsclient -c -e '(make-vterm-frame)'"), desc = "Launch VTerm"),
                           Key (["shift"], "q", lazy.spawn ("emacs -Q"), desc = "Launch Emacs without loading configs")],
              name = "Emacs"),
    # Toggle between different layouts as defined below
    Key ([mod], "Tab", lazy.next_layout(), desc="Toggle between layouts"),
    Key ([mod, "shift"], "Tab", lazy.prev_layout(), desc="Toggle between layouts"),
    Key ([mod, "shift"], "q", lazy.window.kill(), desc="Kill focused window"),
    Key ([mod, "shift"], "r", lazy.restart(), desc="Restart Qtile"),
    Key ([mod, "control"], "q", lazy.shutdown(), desc="Shutdown Qtile"),
    Key ([mod], "d", lazy.run_extension (extension.DmenuRun (dmenu_prompt = ">", dmenu_font = "IBMPlexMono-14", selected_background = "#478062"))),
    Key ([mod, "shift"], "space", lazy.window.toggle_floating (), desc = "Toggle between floating"),
    Key ([mod], "f", lazy.window.toggle_fullscreen (), desc = "Toggle fullscreen"),
    Key ([mod], "Print", lazy.spawn ("xfce4-screenshooter -cr"), desc = "Take a region screeshot and copy to clipboard"),
    Key ([mod, "shift"], "Print", lazy.spawn ("xfce4-screenshooter"), desc = "Open xfce4-screenshooter"),
    Key ([mod, "control"], "Print", lazy.spawn ("xfce4-screenshooter -fc"), desc = "Take a full screenshot and copy to clipboard"),
    KeyChord ([mod], "x", [Key ([], "x", lazy.spawn ("i3lock -i /home/id/dotfiles/wallpapers/void_linux_1_1080.png"), desc = "Lock the screen"),
                           Key ([], "p", lazy.spawn ("pavucontrol"), desc = "Spawn pavucontrol"),
                           Key ([], "t", lazy.spawn ("thunar"), desc = "Launch thunar"),],
              name = "Utilities"),
]

keys.extend ([
    Key ([mod], "k", lazy.function (traverse.up)),
    Key ([mod], "j", lazy.function (traverse.down)),
    Key ([mod], "h", lazy.function (traverse.left)),
    Key ([mod], "l", lazy.function (traverse.right)),
    ])

groups = [ScratchPad ("scratchpad", [DropDown ("term", "alacritty -t scratchpad", height = 0.5, width = 0.5,
                                               x = 0.25, y = 0.25, opacity = 1),
                                     DropDown ("pavu", "pavucontrol", height = 0.5, width = 0.5, x = 0.25,
                                               y = 0.25, opacity = 1),
                                     DropDown ("keepass", "/home/id/dotfiles/scripts/password-dropdown.sh",
                                               height = 0.5, width = 0.5, x = 0.25, y = 0.25,
                                               opacity = 1, match = Match (wm_class = "KeePassXC")),
                                     DropDown ("emacs", "emacsclient -c -a ''", height = 0.5, width = 0.5,
                                               x = 0.25, y = 0.25, opacity = 1, match = Match (wm_class = "Emacs")),
                                     DropDown ("calc", "emacsclient -c -e '(calc nil t)'", height = 0.5,
                                               width = 0.5, x = 0.25, y = 0.25, opacity = 1,
                                               match = Match (wm_class = "Emacs")),
                                     DropDown ("vterm", "emacsclient -c -e '(make-vterm-frame)'", height = 0.5,
                                               width = 0.5, x = 0.25, y = 0.25, opacity = 1,
                                               match = Match (wm_class = "Emacs")),]),
          Group (name = "1", label = "1: "),
          Group (name = "2", label = "2: ", matches = [Match (wm_class = "discord")]),
          Group (name = "3", label = "3: "),
          Group (name = "4", label = "4: "),
          Group (name = "5"),
          Group (name = "6"),
          Group (name = "7"),
          Group (name = "8"),
          Group (name = "9", matches = [Match (wm_class = "Steam")]),
          Group (name = "0"),]

for i in groups[1:]:
    keys.extend(
        [
            # mod1 + letter of group = switch to group
            Key(
                [mod],
                i.name,
                lazy.group[i.name].toscreen(),
                desc="Switch to group {}".format(i.name),
            ),
            # mod1 + shift + letter of group = switch to & move focused window to group
            Key(
                [mod, "shift"],
                i.name,
                lazy.window.togroup(i.name, switch_group=True),
                desc="Switch to & move focused window to group {}".format(i.name),
            ),
            # Or, use below if you prefer not to switch to that group.
            # # mod1 + shift + letter of group = move focused window to group
            # Key([mod, "shift"], i.name, lazy.window.togroup(i.name),
            #     desc="move focused window to group {}".format(i.name)),
        ]
    )

keys.extend ([KeyChord ([mod], "grave", [Key ([], "Return", lazy.group["scratchpad"].dropdown_toggle ("term")),
                                         Key ([], "p", lazy.group["scratchpad"].dropdown_toggle ("pavu")),
                                         Key ([], "k", lazy.group["scratchpad"].dropdown_toggle ("keepass")),
                                         Key ([], "e", lazy.group["scratchpad"].dropdown_toggle ("emacs")),
                                         Key ([], "c", lazy.group["scratchpad"].dropdown_toggle ("calc")),
                                         Key ([], "v", lazy.group["scratchpad"].dropdown_toggle ("vterm")),],
                        name = "Scratchpad")])

layouts = [
    layout.Spiral (margin = 4),
    # layout.Bsp(border_width=2, margin = 4, fair = False),
    layout.Max(),
    layout.Columns(border_focus_stack=["#d75f5f", "#8f3d3d"], border_width=2, margin = 4, insert_position = 1),
    # Try more layouts by unleashing below layouts.
    # layout.Stack(num_stacks=2),
    # layout.Matrix(),
    # layout.MonadTall(),
    # layout.MonadWide(),
    # layout.RatioTile(),
    # layout.Tile(),
    # layout.TreeTab(),
    # layout.VerticalTile(),
    # layout.Zoomy(),
]

widget_defaults = dict(
    font="IBMPlexMono",
    fontsize=14,
    padding=3,
)
extension_defaults = widget_defaults.copy()

screens = [
    Screen(
        top=bar.Bar(
            [
                widget.GroupBox(highlight_method = "block", hide_unused = True, this_current_screen_border = "#478062", margin_x = 0, rounded = False),
                widget.Spacer (length = 5),
                widget.CurrentLayout(fmt = "[{}]"),
                widget.Spacer (length = 5),
                # widget.Prompt(),
                widget.WindowName(background = "#478062"),
                # widget.Chord(
                #     chords_colors={
                #         "launch": ("#ff0000", "#ffffff"),
                #     },
                #     name_transform=lambda name: name.upper(),
                # ),
                widget.Spacer (length = 5),
                widget.Chord (background = "#AAC2AB", foreground = "#000000"),
                widget.CPU (format = "[CPU {load_percent}%]"),
                widget.Spacer (length = 5),
                widget.Memory (format = "[RAM {MemUsed:.1f}GiB]", measure_mem = "G"),
                widget.Spacer (length = 5),
                widget.Clock(format="%Y-%m-%d %a %H:%M"),
                widget.Spacer (length = 5),
                widget.Systray(),
            ],
            25,
            # border_width=[2, 0, 2, 0],  # Draw top and bottom borders
            # border_color=["ff00ff", "000000", "ff00ff", "000000"]  # Borders are magenta
        ),
        wallpaper = "~/dotfiles/wallpapers/void_linux_chan.png",
        wallpaper_mode = "fill",
    ),
]

# Drag floating layouts.
mouse = [
    Drag([mod], "Button1", lazy.window.set_position_floating(), start=lazy.window.get_position()),
    Drag([mod], "Button3", lazy.window.set_size_floating(), start=lazy.window.get_size()),
    Click([mod], "Button2", lazy.window.bring_to_front()),
]

dgroups_key_binder = None
dgroups_app_rules = []  # type: list
follow_mouse_focus = False
bring_front_click = False
cursor_warp = False
floating_layout = layout.Floating(
    float_rules=[
        # Run the utility of `xprop` to see the wm class and name of an X client.
        *layout.Floating.default_float_rules[:-1],
        Match(wm_class="confirmreset"),  # gitk
        Match(wm_class="makebranch"),  # gitk
        Match(wm_class="maketag"),  # gitk
        Match(wm_class="ssh-askpass"),  # ssh-askpass
        Match(title="branchdialog"),  # gitk
        Match(title="pinentry"),  # GPG key password entry
        Match (title = "About Mozilla Firefox"),
        Match (title = "em-float"),
    ]
)
auto_fullscreen = True
focus_on_window_activation = "smart"
reconfigure_screens = True

# If things like steam games want to auto-minimize themselves when losing
# focus, should we respect this or not?
auto_minimize = True

# When using the Wayland backend, this can be used to configure input devices.
wl_input_rules = None

# XXX: Gasp! We're lying here. In fact, nobody really uses or cares about this
# string besides java UI toolkits; you can see several discussions on the
# mailing lists, GitHub issues, and other WM documentation that suggest setting
# this string if your java app doesn't work correctly. We may as well just lie
# and say that we're a working one by default.
#
# We choose LG3D to maximize irony: it is a 3D non-reparenting WM written in
# java that happens to be on java's whitelist.
wmname = "LG3D"

@hook.subscribe.startup_once
def autostart_once ():
    sh = os.path.expanduser ("~/dotfiles/misc/qtile_startup_once.sh")
    subprocess.run (sh)

@hook.subscribe.shutdown
def qtile_shutdown ():
    subprocess.run ("emacsclient", "-e", "'(save-buffers-kill-emacs)'")

@lazy.function
def float_to_front (qtile):
    for group in qtile.groups:
        for window in group.windows:
            if window.floating:
                window.cmd_bring_to_front()    
keys.append (Key ([mod, "shift"], "f", float_to_front (), desc = "bring all floating windows to front"))
